src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs

```diff
@@ -330,59 +330,71 @@ internal Task ExtractRelativeToDirectoryAsync(string destinationDirectoryPath, b
             {
-                throw new IOException(SR.Format(SR.TarExtractingResultsFileOutside, Name, destinationDirectoryPath));
+                throw new IOException(SR.Format(SR.TarExtractingResultsFileOutside, name, destinationDirectoryPath));
             }
 
             string? linkTargetPath = null;
-            if (EntryType is TarEntryType.SymbolicLink or TarEntryType.HardLink)
+            if (EntryType is TarEntryType.SymbolicLink)
             {
-                if (string.IsNullOrEmpty(LinkName))
+                // LinkName is an absolute path, or path relative to the fileDestinationPath directory.
+                string linkName = ArchivingUtils.SanitizeEntryFilePath(LinkName, preserveDriveRoot: true);
+                if (linkName.Length == 0)
                 {
                     throw new InvalidDataException(SR.TarEntryHardLinkOrSymlinkLinkNameEmpty);
                 }
 
-                linkTargetPath = GetSanitizedFullPath(destinationDirectoryPath,
-                    Path.IsPathFullyQualified(LinkName) ? LinkName : Path.Join(Path.GetDirectoryName(fileDestinationPath), LinkName));
-
+                // Disallow link targets that point outside the destination directory.
+                linkTargetPath = GetFullDestinationPath(
+                                    destinationDirectoryPath,
+                                    Path.IsPathFullyQualified(linkName) ? linkName : Path.Join(Path.GetDirectoryName(fileDestinationPath), linkName));
                 if (linkTargetPath == null)
                 {
-                    throw new IOException(SR.Format(SR.TarExtractingResultsLinkOutside, LinkName, destinationDirectoryPath));
+                    throw new IOException(SR.Format(SR.TarExtractingResultsLinkOutside, linkName, destinationDirectoryPath));
+                }
+                linkTargetPath = linkName;
+            }
+            else if (EntryType is TarEntryType.HardLink)
+            {
+                // LinkName is path relative to the destinationDirectoryPath.
+                string linkName = ArchivingUtils.SanitizeEntryFilePath(LinkName, preserveDriveRoot: false);
+                if (linkName.Length == 0)
+                {
+                    throw new InvalidDataException(SR.TarEntryHardLinkOrSymlinkLinkNameEmpty);
                 }
 
-                // after TarExtractingResultsLinkOutside validation, preserve the original
-                // symlink target path (to match behavior of other utilities).
-                linkTargetPath = LinkName;
+                // Disallow link targets that point outside the destination directory.
+                linkTargetPath = GetFullDestinationPath(destinationDirectoryPath, Path.Join(destinationDirectoryPath, linkName));
+                if (linkTargetPath == null)
+                {
+                    throw new IOException(SR.Format(SR.TarExtractingResultsLinkOutside, linkName, destinationDirectoryPath));
+                }
```

Comment:
What do we do if the hard link target does not exist at the moment of extraction i.e: on `CreateNonRegularFile`? I expect that the respective syscall fails.

Reply:
Yes it fails, and that is the proper thing to do.
Hard links are detected and handled by the tar writer. It won't include a hard link before the target file.

note: .NET's tar writer doesn't detect hard links and treats them like regular files. There is an open issue to handle them: https://github.com/dotnet/runtime/issues/74404.